module4-authentication-and-security/r1-securing-backend-app/README.md

```diff
@@ -0,0 +1,141 @@
+
+To make this a bit harder to brute force, you can create a token system. Tokens are random strings of characters, usually with arbitrary length, that are intended to be hard to brute force (guess). A token can look something like this: `4f57ec27561d4524e6b5bb1f08fb7cccfb36a5e5` and is generated using special functions called _Cryptographically Secure Pseudo-Random Number Generator_ (CSPRNG). For example, you can run this code inside NodeJS to get a secure token using the `crypto` NodeJS library:
+
+```js
+const crypto = require("crypto");
+const size = 20; // bytes (or 40 hex digits)
+
+const token = crypto.randomBytes(size).toString("hex");
+console.log(token); // example output: 5d4225087676cf60670144e37b1694355a2e24d2
+```
+
+You can change the column `user_email` to `token` and every time a user comes in, you can ask them for the token, or allow them to generate a token. If they are new users, they can generate a new token, and save it, then start creating their lists. When they visit again, they enter that token to get their lists back.
+
+![Bingo!](https://media.giphy.com/media/W5dBnWmcGDABi/giphy.gif)
+
+Bingo! You've just created your first, very adequate authentication layer! You've secured your app and **slimmed the chances** of brute-forcing user data! Many apps today use this pattern to secure data. However, keep in mind that you usually need other features to track, like user email, name, age, etc. So such a simplistic approach wouldn't suffice. That's when this token is moved into another table and tracked with a `user_id` instead.
+
+## Why do we need passwords
+
+The token system developed above is very robust, but it isn't always an acceptable security measure (at least not alone). That's because, if a user loses the token, it is next to impossible (especially if it was generated with CSPRNGs with reasonable size) to gain access back to their lists using today's technologies.
+
+So, maybe instead of a long random token, use a password? We can ask the user when they land on our page for a passcode (aka password). And we check the database if we can find a list where `password = USER_SUBMITTED_PASSCODE` then display them. If not, that means they are a new user.
+
+Indeed this works, passwords are okay and a perfectly viable and widespread approach. Keeping in mind that users tend to use easy to remember words, dates, phone numbers, and the like. This makes passwords vulnerable to brute-force attacks. It would be relatively harder than brute-forcing emails, but still easy to guess, and very easy for computers to break by checking every possible combination of words, dates, numbers, and even relatively harder passwords using a technology called **rainbow tables**.
+
+Hence, there are two things to remember about passwords:
+They require validation to prevent brute force attacks (thus preventing easy to guess passwords)
```

Comment:
use bullet pts

Reply:
also server-side validation such as captcha or throttling IPs is how a lot of big sites are going to prevent you from brute forcing a users email. for example go to google and guess the wrong pw 2x, you're gonna get captcha'd